File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**9)

preorder = []
while True:
    try:
        v = int(input())
        preorder.append(v)
    except:
        break

def get_preorder(start, end):
    if start > end:
        return
    
    node = preorder[start]
    mid = end + 1
    for i in range(start+1, end+1):
        if node < preorder[i]:
            mid = i
            break
    
    get_preorder(start+1, mid-1)
    get_preorder(mid, end)
    print(node)

get_preorder(0, len(preorder)-1)

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .trim();

const pre_traversal = input.split("\n").map(Number);

function visit_tree(start, end) {
  if (end < start) return;

  let mid = end + 1;
  for (let i = start + 1; i <= end; i++) {
    if (pre_traversal[start] < pre_traversal[i]) {
      mid = i;
      break;
    }
  }

  visit_tree(start + 1, mid - 1);
  visit_tree(mid, end);

  console.log(pre_traversal[start]);
}

visit_tree(0, pre_traversal.length - 1);
